### ISSUE TITLE
convert to onnx

### ISSUE THREAD

**Issue by qiuzhewei** [oufuzhao/CLIB-FIQA#5]
Hello,
Is there any script available to convert pytorch model to ONNX?
Thanks in advance!

**Comment by oufuzhao** [comment]
Hello Zhewei,

Thanks for your interest in our work.

The following is the code that I used to convert the pytorch model to onnx, but I tried to install the CNN model before, not the CLIP model. You can try to modify the following code to convert it:


---------
import io
import torch
import torch.onnx
import my_model                                                         # My model python file


def convert(pthfile, onnx_model_name):
  model = my_model().to('cpu')                                                              # pls modity - module name
  net_dict = model.state_dict()  
  pretrained_dict = torch.load(pthfile, map_location='cpu')
  pretrained_dict =  {k.replace('module.',''): v for k, v in pretrained_dict.items()}
  same_dict =  {k: v for k, v in pretrained_dict.items() if k in net_dict}
  diff_dict = {k: v for k, v in net_dict.items() if k not in pretrained_dict}
  net_dict.update(same_dict)
  model.load_state_dict(net_dict)
  dummy_input1 = torch.randn(1, 3, 224, 224)                                                # input size to test
  input_names = ["actual_input_1"]
  output_names = ["output1"]
  torch.onnx.export(model, dummy_input1, onnx_model_name, verbose=True, input_names=input_names, output_names=output_names)      # pls modity - ONNX model name ("FIQA_model.onnx")
 
if __name__ == "__main__":
  onnx_model_name = "FIQA_model.onnx"
  pthfile = 'xxxxxxxx/xxxxxxxxxx/xxxxxxxx.pth'               # pls modity - path of pth file 
  convert(pthfile, onnx_model_name)

**Comment by qiuzhewei** [comment]
Thanks for your replay. Succeeded in converting the model. Below is a copy of script in case anyone needs it.

```
import torch
import torch.onnx
from inference import backboneSet
from utilities import load_net_param, convert_models_to_fp32
from itertools import product
from model import clip

quality_list = ['bad', 'poor', 'fair', 'good', 'perfect']
blur_list = ['hazy', 'blurry', 'clear']
occ_list = ['obstructed', 'unobstructed']
pose_list = ['profile', 'slight angle', 'frontal']
exp_list = ['exaggerated expression', 'typical expression']
ill_list = ['extreme lighting', 'normal lighting']
joint_texts = torch.cat([clip.tokenize(f"a photo of a {b}, {o}, and {p} face with {e} under {l}, which is of {q} quality") 
                for b, o, p, e, l, q in product(blur_list, occ_list, pose_list, exp_list, ill_list, quality_list)]).cpu()


def convert(pthfile, netfile, onnx_model_name, fp32=True):
    model = backboneSet(pthfile).to('cpu') # pls modity - module name
    model = load_net_param(model, netfile)
    if fp32:
        convert_models_to_fp32(model)
    dummy_input1 = torch.randn(1, 3, 224, 224).cpu() # input size to test
    dummy_input2 = joint_texts
    input_names = ["input1", "input2"]
    output_names = ["output1"]
    torch.onnx.export(model, 
                      (dummy_input1, dummy_input2), 
                      onnx_model_name, 
                      verbose=True,
                      opset_version=14,
                      input_names=input_names, 
                      output_names=output_names) # pls modity - ONNX model name ("FIQA_model.onnx")

if __name__ == "__main__":
    onnx_model_name = "FIQA_model.onnx"
    pthfile = "./weights/RN50.pt" # pls modity - path of pth file
    netfile = "./weights/CLIB-FIQA_R50.pth"
    convert(pthfile, netfile, onnx_model_name)
```

**Comment by qiuzhewei** [comment]
And one more question, I notice that for any score, e.g. occlusion, after summation over various axis, the output seems to be either obstructed or unobstructed with a high score, very close to 1, [2.1421205087523987e-11, 1.0] for example, rarely see any intermediate score, [0.4, 0.6] for example.
This is kind of weird, any clue about that?

**Comment by oufuzhao** [comment]
Hi Zhewei,

Thanks for sharing the conversion code to help people in need.

Honestly, I am not sure about the issue with extreme output probabilities, but I think it may be related to the imbalance of multi-label classifications during training. As our main task is to fit the quality distribution, we did not pay much attention to the prediction performance in auxiliary classification tasks.
